### PR TITLE
simplification of the Application::bootstrap method

### PR DESCRIPTION
**Pull request by Faryshta**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | https://scrutinizer-ci.com/g/yiisoft/yii2/code-structure/master/operation/yii%5Cbase%5CApplication%3A%3Abootstrap

### CODE REVIEW

**Comment by Faryshta** [comment]
the coverage is also very poor and none of the lines I edited gets covered by unit testing. I couldn't find any test related to bootstrap or Application

File: framework/base/Application.php [hunk 1]
```diff
@@ -279,44 +279,73 @@ protected function bootstrap()
...
-                } elseif ($this->hasModule($class)) {
-                    $component = $this->getModule($class);
-                } elseif (strpos($class, '\\') === false) {
-                    throw new InvalidConfigException("Unknown bootstrapping component ID: $class");
-                }
-            }
-            if (!isset($component)) {
-                $component = Yii::createObject($class);
-            }
+            $this->bootstrapComponent($this->getComponent($class));
```
Comment (samdark) [inline framework/base/Application.php]: I'm not sure how a function call in foreach will affect overall performance...

Comment (samdark) [inline framework/base/Application.php]: Would be great to do some tests.

**Comment by samdark** [comment]
From the refactoring standpoint it makes sense.

**Comment by dynasource** [comment]
we first have to have proper unit tests before we can safely refactor

**Comment by samdark** [comment]
Agree.

**Comment by Faryshta** [comment]
Agree. But I don't know how to make such tests, specially since everytime you create a new `Application` object the variable `Yii::$app` gets rewritten

**Comment by SilverFire** [comment]
@Faryshta you can call `mockApplication` for each particular test

```php
public function testSomeAnnoyingCrap()
{
       $app = $this->mockApplication([
            'bootstrap' => [
                // ...
            ]
       ]);

       // go ahead
}
```

**Comment by dynasource** [comment]
Im preparing basic tests for Application. Will do a PR within a few days,

**Comment by Faryshta** [comment]
@dynasource could you work on it?

@SilverFire but then how would i make the assertions?

**Comment by dynasource** [comment]
Im quite busy this month. Will do in May.

**Comment by cebe** [comment]
In current state this PR can not be merged and it is outdated. Closing this for now. It also adds unnecessary methods to public interface, which I see no good reasons for.

**Pull request closed without merging**